crates/ra_hir/src/code_model.rs

```diff
@@ -960,6 +960,38 @@ impl ImplDef {
         db.impl_data(self.id).target_trait.clone()
     }
 
+    pub fn target_trait_substs_matches(&self, db: &dyn HirDatabase, typs: &[Type]) -> bool {
+        let type_ref = match self.target_trait(db) {
+            Some(typ_ref) => typ_ref,
+            None => return false,
+        };
+        let resolver = self.id.resolver(db.upcast());
+        let ctx = hir_ty::TyLoweringContext::new(db, &resolver);
+        let ty = Ty::from_hir(&ctx, &type_ref);
+        let d = match ty.dyn_trait_ref() {
+            Some(d) => d,
+            None => return false,
+        };
+        let mut matches = true;
+        let mut i = 0;
+        d.substs.walk(&mut |t| {
```

Comment:
I don't think you need to use `walk` here; that does a deep walk, recursing into the types. Just `d.substs.iter()` should suffice. Also note that the first type parameter for the trait reference is the implemented self type: so the whole trait ref isn't just `From<Foo>`, it's `Bar: From<Foo>`. So in your case, you can check the implemented target type with this call as well.